language/stdlib/modules/Errors.move

```diff
@@ -0,0 +1,75 @@
+address 0x1 {
+
+/// Module defining error codes used in Move aborts throughout the framework.
+///
+/// A `u64` error code is constructed from two values:
+///
+///  1. The *error category* which is encoded in the lower 8 bits of the code. Error categories are
+///     declared in this module and are globally unique across the Libra framework. There is a limited
+///     fixed set of predefined categories, and the framework is guaranteed to use those consistently.
+///
+///  2. The *error reason* which is encoded in the remaining 54 bits of the code. The reason is a unique
+///     number relative to the module which raised the error and can be used to obtain more information about
+///     the error at hand. It is mostly used for diagnosis purposes. Error reasons may change over time as the
+///     framework evolves. TODO(wrwg): determine what kind of stability guarantees we give about reasons/
+///     associated module.
+module Errors {
+
+    /// A function to create an error from from a category and a reason.
+    fun make(category: u8, reason: u64): u64 {
+        (category as u64) + (reason << 8)
+    }
+
+    /// The system is in a state where the performed operation is not allowed. Example: call to a function only allowed
+    /// in genesis.
+    const INVALID_STATE: u8 = 1;
+
+    /// The signer of a transaction has not the expected address for this operation. Example: a call to a function which
+    /// requires a signer which is the Libra root.
```

Comment:
This might be a confusing example because Libra Root is a role (and thus the reader might wonder "why not use `REQUIRES_ROLE` here?"). Maybe something like "a function that publishes a singleton resource under a particular address" could work.

Reply:
Done